gene_expression_processing.py

```diff
@@ -0,0 +1,272 @@
+
+
+@click.command()
+@click.option('--celltypes')
+@click.option('--chromosomes')
+@click.option('--gene-info-tsv')
+@click.option('--expression-files-prefix')
+@click.option('--sample-mapping-file-path')
+@click.option('--min-pct-expr')
+@click.option('--cis-window-size')
+@click.option(
+    '--max-gene-concurrency',
+    type=int,
+    default=50,
+    help=(
+        'To avoid resource starvation, set this concurrency to limit '
+        'horizontal scale. Higher numbers have a better walltime, but '
+        'risk jobs that are stuck (which are expensive)'
+    ),
+)
+def expression_pipeline(
+    celltypes: str,
+    chromosomes: str,
+    gene_info_tsv: str,
+    expression_files_prefix: str,
+    sample_mapping_file_path: str,
+    min_pct_expr: int = 5,
+    cis_window_size: int = 100000,
+    max_gene_concurrency=100,
+):
+    """
+    Run expression processing pipeline
+    """
+
+    logging.info(f'Cell types to run: {celltypes}')
+    logging.info(f'Chromosomes to run: {chromosomes}')
+
+    # create phenotype covariate files
+    smf_df = pd.read_csv(sample_mapping_file_path, sep='\t')
+    gene_info_df = pd.read_csv(gene_info_tsv, sep='\t')
+
+    # Setup MAX concurrency by genes
+    _dependent_jobs: list[hb.batch.job.Job] = []
+
+    def manage_concurrency_for_job(job: hb.batch.job.Job):
+        """
+        To avoid having too many jobs running at once, we have to limit concurrency.
+        """
+        if len(_dependent_jobs) >= max_gene_concurrency:
+            job.depends_on(_dependent_jobs[-max_gene_concurrency])
+        _dependent_jobs.append(job)
+
+    for celltype in celltypes.split(','):
+        # get covariates (cell type specific)
+        cov_df = get_celltype_covariates(
+            expression_files_prefix=expression_files_prefix, cell_type=celltype
+        )
+        for chromosome in chromosomes.split(','):
+            # get expression (cell type + chromosome)
+            expr_adata: sc.AnnData = get_chrom_celltype_expression(
+                gene_info_df=gene_info_df,
+                expression_files_prefix=expression_files_prefix,
+                chromosome=chromosome,
+                cell_type=celltype,
+            )
+            # remove lowly expressed genes
+            filter_adata: sc.AnnData = filter_lowly_expressed_genes(
+                expression_adata=expr_adata, min_pct=min_pct_expr
+            )
+
+            # combine files for each gene
+            # pylint: disable=no-member
+            for gene in filter_adata.raw.var.index:
+                pheno_cov_job = get_batch().new_python_job(name='creta pheno cov job')
+                copy_common_env(pheno_cov_job)
+                pheno_cov_job.image(CELLREGMAP_IMAGE)
+
+                # pass the output file path to the job, don't expect an object back
+                pheno_cov_job.call(
+                    build_pheno_cov_filename,
+                    gene_name=gene,
+                    cov_df=cov_df,
```

Comment:
Passing in dataframes here is probably a bad idea, it does the serialisation / deserialisation step in pickle, which is likely significantly slower than a regular write + read.